music/src/music_vae/midi_me.ts

```diff
@@ -0,0 +1,299 @@
+  constructor() {
+    super({});
+  }
+
+  computeOutputShape(inputShape: tf.Shape[]) {
+    return inputShape[0];
+  }
+
+  call(inputs: [tf.Tensor2D, tf.Tensor2D]) {
+    return tf.tidy(() => {
+      const [mu, sigma] = inputs;
+      return tf.add(tf.mul(tf.randomNormal(sigma.shape), sigma), mu);
+    });
+  }
+  getClassName() {
+    return 'SamplingLayer';
+  }
+}
+
+/**
+ * An interface for providing configurable properties to the MidiMe model.
+ * @param input_size The shape of the VAE input. Since the inputs to this
+ * VAE are actually latent vectors from MusicVAE, then this number should be
+ * equal to the number of latent variables used by MusicVAE (`zDims`). The
+ * default is 256.
+ * @param output_size The size of the output. The default is 4.
+ * @param encoder_layers The shape of the layers in the Encoder network.
+ * @param decoder_layers The shape of the layers in the Decoder network.
+ * @param beta Weight of the latent loss in the total VAE loss. Default is 1.
+ * @param epochs Number of epochs to train for. Default is 10.
+ */
+interface MidiMeConfig {
+  input_size?: number;
+  output_size?: number;
+  encoder_layers?: number[];
+  decoder_layers?: number[];
+  beta?: number;
+  // For training:
+  epochs?: number;
+}
+
+/**
+ * Main `MidiMe` model class.
+ *
+ * A `MidiMe` is a hierarchical variational autoencoder that is trained on
+ * latent vectors generated by `MusicVAE`. It allows you to personalize your own
+ * MusicVAE model with just a little data, so that samples from MidiMe sound
+ * similar to the input data.
+ */
+class MidiMe {
+  // Model configuration.
+  public config: MidiMeConfig;
+
+  // Main model and submodels.
+  private vae: tf.LayersModel;
+  private encoder: tf.LayersModel;
+  private decoder: tf.LayersModel;
+
+  trained = false;
+  initialized = false;
+
+  /**
+   * `MidiMe` constructor.
+   *
+   * @param config (optional) Model configuration.
+   */
+  constructor(config: MidiMeConfig = {}) {
+    this.config = {
+      encoder_layers: config.encoder_layers || [1024, 256, 64],
+      decoder_layers: config.decoder_layers || [64, 256, 1024],
+      input_size: config.input_size || 256,
+      output_size: config.output_size || 4,
+      beta: config.beta || 1,
+      epochs: config.epochs || 10
+    };
+  }
+
+  /**
+   * Disposes of any untracked tensors to avoid GPU memory leaks.
+   */
+  dispose() {
+    if (!this.initialized) {
+      return;
+    }
+    this.encoder.dispose();
+    this.decoder.dispose();
+    this.vae.dispose();
+    this.initialized = false;
+  }
+
+  /**
+   * Instantiates the `Encoder`, `Decoder` and the main `VAE`.
+   */
+  initialize() {
+    this.dispose();
+
+    const startTime = performance.now();
+    const z = tf.input({shape: [this.config['input_size']]});
+
+    // Encoder model, goes from the original input, returns an output.
+    this.encoder = this.getEncoder(z);
+    const [sampleZ, , ] = this.encoder.apply(z) as tf.SymbolicTensor[];
```

Comment:
Maybe we should call sampleZ something like zPrime? It should be more clear that this is the latent vector for the MidiMe VAE.

Reply:
Actually, let's just rename `z` -> `x`